src/components/cylc/tree/tree.js

```diff
@@ -0,0 +1,497 @@
+ * having to iterate and visit each of its parents.
+ *
+ * <script language="javascript">
+ * const jobID = 'kinow|five|20150129T1200+12|foo|1'
+ * const someJob = cycleTreeInstance.lookup.get(jobID)
+ * </script>
+ *
+ * Finally, this data structure class contains methods to:
+ *
+ * - add a node of any type
+ * - update a node of any type
+ * - remove a node of any type
+ *
+ * When a node is added, it gets added to the hierarchical tree (i.e. it
+ * will be added as child of some other node), and also added to the
+ * lookup map.
+ *
+ * When a node is updated, the values of the given node argument in the
+ * function replace the values in the existing element of the map. i.e.
+ * we will find the object in the lookup map, and use Lodash's `merge`.
+ *
+ * And when a node is removed, it gets removed from its parent's `.children`
+ * array, and the node and each of its children get removed from the
+ * lookup map as well.
+ */
+class CylcTree {
+  /**
+   * Create a tree with an initial root node, representing
+   * a workflow in Cylc.
+   *
+   * @param {null | {
+   *   id: string,
+   *   node: Object,
+   *   children: []
+   * }} [workflow]
+   */
+  constructor (workflow) {
+    /**
+     * @type {Map<string, any>}
+     */
+    this.lookup = new Map()
+    if (!workflow) {
+      this.root = {
+        id: '',
+        node: {},
+        children: []
+      }
+    } else {
+      this.root = workflow
+      this.lookup.set(this.root.id, this.root)
+    }
+  }
+
+  setWorkflow (workflow) {
+    if (!workflow) {
+      throw new Error('You must provide a valid workflow!')
+    }
+    this.root = workflow
+    this.lookup.set(workflow.id, workflow)
+  }
+
+  clear () {
+    this.lookup.clear()
+    this.root = {
+      id: '',
+      node: {},
+      children: []
+    }
+  }
+
+  isEmpty () {
+    return this.lookup.size === 0
+  }
+
+  /**
+   * @param {null|{
+   *   id: string,
+   *   node: Object,
+   *   children: []
+   * }} node
+   */
+  recursivelyRemoveNode (node) {
+    if (node) {
+      const stack = [node]
+      while (stack.length > 0) {
+        const n = stack.pop()
+        this.lookup.delete(n.id)
+        if (n.children && n.children.length > 0) {
+          stack.push(...n.children)
+        }
+      }
+    }
+  }
+
+  // --- Cycle points
+
+  /**
+   * @param {null|{
+   *   id: string,
+   *   node: Object,
+   *   children: []
+   * }} cyclePoint
+   */
+  addCyclePoint (cyclePoint) {
+    if (cyclePoint && !this.lookup.has(cyclePoint.id)) {
+      this.lookup.set(cyclePoint.id, cyclePoint)
+      this.root.children.push(cyclePoint)
+      // sort cycle points
+      this.root.children.sort((cyclepoint, anotherCyclepoint) => {
+        return cyclepoint.id.localeCompare(anotherCyclepoint.id)
+      })
+    }
+  }
+
+  /**
+   * @param {null|{
+   *   id: string,
+   *   node: Object,
+   *   children: []
+   * }} cyclePoint
+   */
+  updateCyclePoint (cyclePoint) {
+    if (cyclePoint) {
```

Comment:
Out of interest why do we need to defend against `null` values in each method?

Reply:
I wrote this code before the deltas. I actually re-used a lot of the existing code that we had in the Vuex store or that was used by the tree Vuex module (not the best place to keep it).

And I wasn't sure what to expect from the deltas, so ended up coding defensively from the beginning.

Now I am not so sure it is necessary. I **think** we could get an undefined if I try to access a key that doesn't exist in the deltas, though that's hard. We should be able to remove this if you think that's better.

Just a matter of removing these `if` statements, removing the tests that were checking for `null`, running unit + e2e tests, and playing a bit with `five`/`families2`/`complex`. A 30-60 minutes job.

Reply:
Thought so, if not necessary any more it would be nice to remove at some point just to streamline the code but no biggie.

Reply:
Done! Tested a couple workflows and it appeared to have worked. We can re-add it later if necessary :+1: